src/services/rpcService.ts

```diff
@@ -1,134 +1,211 @@
+   */
+  async getSolBalance(publicKey: PublicKey): Promise<bigint> {
+    try {
+      if (!validatePublicKey(publicKey)) {
+        return BigInt(0);
       }
+
+      console.log("Debug: getting SOL balance for", publicKey.toBase58());
+      const balance = await this.connection.getBalance(publicKey);
+      if (balance === 0) {
+        console.warn(`Debug: SOL balance for ${publicKey.toBase58()} is 0`);
+        return BigInt(0);
+      }
+      console.log(`Debug: SOL balance for ${publicKey.toBase58()} is ${balance / 1e9} SOL`);
+      return BigInt(balance);
+    } catch (error) {
+      console.error("Error fetching SOL balance:", error);
+      throw error;
     }
-  
-    /**
-     * Get the SOL balance for a given public key
-     * @param {PublicKey} publicKey - The public key to check the balance for
-     * @returns {Promise<BN>} The SOL balance as a BN
-     */
-    async getSolBalance(publicKey: PublicKey): Promise<BN> {
-      try {
-        if (!validatePublicKey(publicKey)) {
-          return new BN(0);
-        }
-  
-        console.log("Debug: getting SOL balance for", publicKey.toBase58());
-        const balance = await this.connection.getBalance(publicKey);
-        if (balance === 0) {
-          console.warn(`Debug: SOL balance for ${publicKey.toBase58()} is 0`);
-          return new BN(0);
+  }
+
+  /**
+   * Get the SPL token balance for a given public key and mint
+   * @param {PublicKey} publicKey - The public key to check the balance for
+   * @param {PublicKey} mint - The SPL token mint
+   * @param {PublicKey} programId - The SPL token program ID (default: TOKEN_PROGRAM_ID)
+   * @returns {Promise<bigint>} The SPL token balance as a bigint
+   */
+  async getSplBalance(publicKey: PublicKey, mint: PublicKey, programId = TOKEN_PROGRAM_ID): Promise<bigint> {
+    try {
+      const pk = new PublicKey(publicKey);
+      if (!validatePublicKey(pk)) {
+        return BigInt(0);
+      }
+
+      console.log("getting spl balance", mint);
+      const response = await this.connection.getTokenAccountsByOwner(
+        pk,
+        {
+          mint: mint,
+          programId: programId
         }
-        console.log(`Debug: SOL balance for ${publicKey.toBase58()} is ${balance / 1e9} SOL`);
-        return new BN(balance);
-      } catch (error) {
-        console.error("Error fetching SOL balance:", error);
-        throw error;
+      );
+
+      if (response.value.length === 0) {
+        console.log("No token account found for this wallet");
+        return BigInt(0);
       }
+
+      // The balance is stored in the last 8 bytes of the account data
+      // The bytes are stored in little-endian format
+      const tokenAccountInfo = response.value[0].account.data;
+      const balance = BigInt('0x' + Buffer.from(tokenAccountInfo.subarray(64, 72)).reverse().toString('hex'));
+
+      console.log("Debug: SPL balance for", pk.toBase58(), "is", balance.toString());
```

Comment:
You can get the token balance with an RPC call:
https://solana.com/docs/rpc/http/gettokenaccountbalance

Reply:
This doesn't support token 2022?

Reply:
^^ actual question. I just don't see any docs around this

Reply:
Looks like it works. read through some old PRs in the rpc client